### PR TITLE
TST: refactor test_read_unicode (parametrize to avoid repetition)

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
This is a follow up to #16524 as suggested by @eerovaher in https://github.com/astropy/astropy/pull/16524#discussion_r1626324856

~based on #16524~

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Review by mhvk — APPROVED** [review]
Looks good to me, but since @eerovaher suggested this, I'll let him have a  look before merging.

File: astropy/utils/tests/test_data.py
Comment (eerovaher) [inline astropy/utils/tests/test_data.py:1151]: No reason not to be strict in tests.
```suggestion
    assert type(contents) is expected_type
```

**Review by eerovaher — COMMENTED** [review]
It would be good if someone with some familiarity with Windows could have a look at this too.

File: astropy/utils/tests/test_data.py
Comment (pllim) [inline astropy/utils/tests/test_data.py:1155]: Eero reminded me of this. Probably also have to make sure this is not a bug in `get_pkg_data_contents`. I will investigate when I get the chance.

Comment (pllim) [reply astropy/utils/tests/test_data.py:1155]: So, is the debate `splitlines()` isn't  necessary? I find that it is still needed on Windows. In fact, the old code had it too so I don't understand what the controversy is. Have to step away. More later.

File: astropy/utils/tests/test_data.py
Comment (pllim) [inline astropy/utils/tests/test_data.py:1154]:
```suggestion
```

I don't think it matters if it is `\n` or `\r\n` when it comes to necessity of splitlines? The old code already had splitlines anyway.

For the record, on Windows 11 with Git Bash, I get `\n` and not `\r\n`. If you want Powershell confirmation, we will have to summon Duy again because I don't want to mess up my env by also installing conda on Powershell.

You can see the endlines like this even without invoke package data. This is what the function is really doing. So you can download `unicode.txt` into a local disk and run the following from `main`:

```python
>>> filename = "unicode.txt"
>>> from astropy.utils.data import get_readable_fileobj
>>> with get_readable_fileobj(filename, encoding="utf-8") as fd:
...     contents = fd.read()
>>> contents  # Oh, it's Hebrew according to Google translate
'האסטרונומי פייתון\n'
```

Comment (neutrinoceros) [reply astropy/utils/tests/test_data.py:1154]: > The old code already had splitlines anyway.

Indeed. This comment is just a testimony of our struggling to remind ourselves why `str.splitlines` is more portable than `str.split('\n')`. I'm not too comfortable with removing it as it would feel like getting back to square one where I (or others) would be tempted to refactor it for whatever reason (I'm reasonably certain I will forget about this oddity again).

Comment (pllim) [reply astropy/utils/tests/test_data.py:1154]: OK then I propose slight verbiage change:

```
    # Using splitlines() instead of split("\n") here for portability as
    # newlines can be represented differently in different OSes.
```

This way, it is clearer why it is considered more portable (from what?) and we don't have to argue how exactly newlines are represented in NIX vs Windows.

Comment (neutrinoceros) [reply astropy/utils/tests/test_data.py:1154]: I agree, that's better ! Thank you !

**Comment by neutrinoceros** [comment]
rebased to fix CI

**Comment by eerovaher** [comment]
I am still interested why does the `\r` show up because [it only appears on Windows (that alone is not surprising) AND when the file is not compressed AND when `encoding="binary"`](https://github.com/astropy/astropy/actions/runs/9372135330/job/25802989914). Does the `/r` get added when the repository is checked out on a Windows machine? Does it get added when Python reads in the file? Does it get added by our `get_pkg_data_contents()` and if it does then is that a bug?

**Comment by neutrinoceros** [comment]
> Does it get added when Python reads in the file?

I don't think that "added" is technically correct but that's what's happening yes. Definitely not within our control and not a bug.

**Comment by pllim** [comment]
I don't think it is just a Python thing. I vaguely remember having to deal with that too back when I tinkered in C. Not sure how much this helps but https://dev.to/pieter/why-windows-uses-rn-newlines-instead-of-n-126l

**Comment by pllim** [comment]
Anyways, this LGTM but I'll let Eero approve in case he still has concerns. Thanks!

**Comment by eerovaher** [comment]
What is the output of
```python
>>> from pathlib import Path
>>> Path("astropy/utils/tests/data/unicode.txt").read_bytes()
```
on Windows?

**Comment by pllim** [comment]
On Windows 11 (Git Bash):

```python
b'# -*- coding: utf-8 -*-\r\n\xd7\x94\xd7\x90\xd7\xa1\xd7\x98\xd7\xa8\xd7\x95\xd7\xa0\xd7\x95\xd7\x9e\xd7\x99 \xd7\xa4\xd7\x99\xd7\x99\xd7\xaa\xd7\x95\xd7\x9f\r\n'
```

**Review by eerovaher — APPROVED** [review]
I see the `\r` also appears in the output of `Path.read_bytes()`, so it's not caused by `get_pkg_data_contents()` and therefore it's not our problem.

Thanks @pllim for trying out these things on Windows so I don't have to.

**Comment by pllim** [comment]
Happy to help. Thanks, all!

**Pull request merged**
